### PR TITLE
introduce ollama processor

### PR DESCRIPTION
**Pull request by rockwotj**
The ollama_completion and ollama_embeddings processor can complete text
and generate embeddings respectively. The use ollama.com server to
perform these operations and with the x_benthos_extra flags will also
embed a server and launch it locally.

I tried to embed the ollama server library directly instead of embedding
the entire server, but it's not quite setup for that at the moment (go
generate is needed in the dependency and I couldn't find a good way to
do that).

Otherwise it's a fairly standard barebones processor, there are more
options for requests and models that can be exposed but I figured that I
would start with the MVP and iterate from there.

### CODE REVIEW

File: internal/impl/ollama/embed_ollama_linux_arm64.go
Comment (rockwotj) [inline internal/impl/ollama/embed_ollama_linux_arm64.go:9]: I re-used this existing extra, but I am happy to either remove the embedded server feature, or use a different tag

File: internal/impl/ollama/embeddings_processor_ollama.go
Comment (mihaitodor) [inline internal/impl/ollama/embeddings_processor_ollama.go:84]: This is a bit annoying, but if you set `[]float64` in here, Bloblang won't be able to operate on it. You'll have to convert it to `[]any`. See docs here: https://pkg.go.dev/github.com/redpanda-data/benthos/v4@v4.32.0/public/service#Message.SetStructured

File: internal/impl/ollama/completion_processor_ollama.go
Comment (mihaitodor) [inline internal/impl/ollama/completion_processor_ollama.go:83]: I'm not entirely sure what's the right thing to do here, but my impression is that if you just return a `service.NewMessage` here, you'll loose any metadata that's set on the origin message and I think most processors don't do that (should look that up, I'm not 100% sure).

**Review by mihaitodor — COMMENTED** [review]
Thanks Tyler! I spotted a few minor issues, but it looks OK otherwise.

File: internal/impl/ollama/base_processor.go
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:71]: If you have ideas on how to hook it up I'm all for doing it - it's actually pretty helpful for troubleshooting.

File: internal/impl/ollama/base_processor.go
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:141]: Good question but no - it's an http.Client under the hood and still subject to network timeouts.

File: internal/impl/ollama/embeddings_processor_ollama.go
Comment (rockwotj) [reply internal/impl/ollama/embeddings_processor_ollama.go:84]: Thank you! 🥇 I just ran into this today with the pinecone output and it was driving me bonkers why I couldn't get the length of this vector

File: internal/impl/ollama/base_processor.go
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:121]: Done

File: .github/workflows/test.yml
Comment (rockwotj) [reply .github/workflows/test.yml:31]: Yes done.

File: internal/impl/ollama/completion_processor_ollama.go
Comment (rockwotj) [reply internal/impl/ollama/completion_processor_ollama.go]: Thanks I looks at a couple other processors (redis, awk) and it seems they preserve metadata as well. I'm good with that - done!

On a side note - should the wasm processor be doing that too? Maybe not because the user is uploading code there and there are mappings for metadata?

File: .github/workflows/test.yml
Comment (rockwotj) [reply .github/workflows/test.yml:31]: I undid this change, redownloading this every `make docs` seems broken, so I've made it more explicit. Maybe there is makefile magic that can prevent running this if the file is already there? Honestly I'm pretty terrible at raw make. Let me know if there a strong preference here!

**Comment by rockwotj** [comment]
FYI you can hold off on reviewing this one, after the batching comments from @mihaitodor I am going to make this a non batch processor and let users specify the batching and concurrently themselves.

**Comment by rockwotj** [comment]
Okay, I've simplified it and we're ready for another looksee here. Also confirmed that we're going to start this one as enterprise for now.

File: internal/impl/ollama/base_processor.go
Comment (mihaitodor) [reply internal/impl/ollama/base_processor.go:70]: Both `proc.Stderr` and `proc.Stdout` are of type [`io.Writer`](https://pkg.go.dev/io#Writer). I'd create a struct which holds a reference to the logger and implements this interface (it just needs a `Write(p []byte) (n int, err error)` method which does a write to the logger member variable from the receiver if that makes sense). Let me know if you need an example.

File: internal/impl/ollama/completion_processor_ollama.go
Comment (mihaitodor) [reply internal/impl/ollama/completion_processor_ollama.go]: It's a tricky question, but if you look [here](https://github.com/redpanda-data/connect/blob/96f6b395e5e2cee71c0bced15ffeb1a575e015ab/internal/impl/wasm/functions.go#L74) and [here](https://github.com/redpanda-data/connect/blob/96f6b395e5e2cee71c0bced15ffeb1a575e015ab/internal/impl/wasm/functions.go#L97), you'll see utilities for reading and setting the metadata from the wasm processor. Users could run a wasm model which overwrites certain metadata fields in the current message, but they might want to preserve the rest. I can imagine that being relevant when processing Kafka messages for example and you may want at the output level to send them to different topics based on the `kafka_topic` metadata field. It would be annoying if the processor pipeline erased the metadata (although you can work around it in various ways, for example by using a `branch`).

**Review by mihaitodor — APPROVED** [review]
Thanks Tyler! I had another look and left some clarifications. Otherwise, it should be OK. I haven't tried to run the code myself, but let me know if you need any help with it.

File: internal/impl/ollama/base_processor.go
Comment (mihaitodor) [reply internal/impl/ollama/base_processor.go:70]: Indeed, it might be more complicated. My initial thinking was that wazero is smart enough to do a single write / log entry, so, even if it contains new lines, we could emit those as-is and leave these as debug logs or, we could also escape the newline characters so we don't end up with multiline logs (I think there's some other stuff which emits multiline logs in debug mode).

Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:70]: Took a stab at this 😄

File: internal/impl/ollama/base_processor.go
Comment (mihaitodor) [inline internal/impl/ollama/base_processor.go:83]: Small nit: I think we might want this path to be configurable in case people deal with readonly file systems.

File: internal/impl/ollama/base_processor.go
Comment (mihaitodor) [reply internal/impl/ollama/base_processor.go]: Awesome, thanks! That looks fine as long as there aren't any concurrent calls to `Write`. Might be worth protecting it with a mutex just in case.

File: .github/workflows/test.yml
Comment (rockwotj) [reply .github/workflows/test.yml]: I cobbled something together - PTAL I'm not good with Makefiles 😄

File: internal/impl/ollama/base_processor.go
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go]: Good call - done!

File: .github/workflows/test.yml
Comment (mihaitodor) [reply .github/workflows/test.yml]: Nice 🥇 That's what I was expecting to work. Thanks for digging into it! I think it's probably not that relevant to have both `ollama-linux-amd64` and `ollama-linux-arm64` or they could be placed in a folder called `ollama` or `ollama-resources` or smth like that so then you won't really need that `if` statement in the Makefile because you can check for the generic name. Guess you can remove the `go generate` stuff from the CGo Dockerfile.

Comment (rockwotj) [reply .github/workflows/test.yml]: Just named them both `ollama` that cleaned things up thanks!

Comment (mihaitodor) [reply .github/workflows/test.yml]: You're welcome! I think you can actually use `./internal/impl/ollama/ollama` as the dependency of the target and then you don't even need the if inside, but that's just a nitpick :)

Comment (rockwotj) [reply .github/workflows/test.yml]: Hrm but it's not working in CI as it can't find the file to embed - it seems to work locally... Any tips?

Comment (mihaitodor) [reply .github/workflows/test.yml]: So this is what I had in mind:
```
./internal/impl/ollama/ollama:
	go generate -tags "$(TAGS)" ./internal/impl/ollama

generate: ./internal/impl/ollama/ollama
```
And then you'd need to make sure `generate` is called first for the `$(APPS)` target like so:

```
$(APPS): %: generate $(PATHINSTBIN)/%
```

That should do the trick I think. If you get stuck let me know and I'll send you the full code.

Comment (rockwotj) [reply .github/workflows/test.yml]: Really appreciate that! That seems to work for me locally, let's see how CI feels about it.

Comment (rockwotj) [reply .github/workflows/test.yml]: Seems to work - thank you! 🙏

Comment (mihaitodor) [reply .github/workflows/test.yml]: You're welcome!

**Review by mihaitodor — APPROVED** [review]
Thanks again Tyler! I don't see any other issues, but I'll let @Jeffail have a look through it as well.

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_chat.adoc:24]:
```suggestion
Sends prompts to your chosen Ollama large language model (LLM) and generates text from the responses, using the Ollama API.
```
I wondered if we could give the user a more detailed description of what the processor does. This is just a suggestion.

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_chat.adoc:26]:
```suggestion
Introduced in version 4.32.0.
```

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_chat.adoc:107]:
```suggestion
The prompt you want to generate a response for. By default, the processor submits the entire payload as a string. This field supports [interpolation functions](http://localhost:8000/configuration/interpolation/#bloblang-queries).
```
Just a suggestion to make this description distinct from the system prompt?

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:24]:
```suggestion
Sends text to your chosen Ollama large language model (LLM) and creates vector embeddings, using the Ollama API. Vector embeddings are long arrays of numbers that represent values or objects, in this case text. 
```
I've suggested a more detailed description using a similar format to the chat processor.

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:36]:
```suggestion
# Common configuration fields, showing default values
```
Brings into line with the style guide.

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:39]: Isn't the local server address the default server address?

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:50]:
```suggestion
# All configuration fields, showing default values
```
Brings into line with the style guide.

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:62]: Do we need this line?

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:64]:
```suggestion
For more information, see the https://ollama.com/[Ollama website^].
```
Just to bring into line with the style guide.

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:70]:
```suggestion
The address of the Ollama server to use. By default, a local Ollama server starts and runs unless you specify the address of a remote server.
```

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:78]:
```suggestion
Unpack the Ollama binary to this directory. Default is `/tmp`.

```

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:86]:
```suggestion
The name of the Ollama LLM to use. For a full list of models, see the https://ollama.com/models[Ollama website].
```

File: docs/modules/components/pages/processors/ollama_embeddings.adoc
Comment (asimms41) [inline docs/modules/components/pages/processors/ollama_embeddings.adoc:106]: The text you want to create vector embeddings for. By default, the processor submits the entire payload as a string.

**Review by asimms41 — CHANGES_REQUESTED** [review]
Sorry, I've just realised that I've added my comments to the .adoc versions of the processors. I'll add them to the Go files for the next set of reviews.

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:36]: This is common code in another place of the codebase, I can open another PR to update.

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:39]: Writing `http://127.0.0.1:11434` is different because then the processor will not start a locally running ollama instance but use an already running server. Any ideas on how to make that clearer?

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:50]: I will open another PR to fix

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:65]: Thank you!

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:63]: I think it will be better with another one of your suggestions

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:71]: Thanks, done!

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:79]: Done, thanks!

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:107]: Thanks, much better!

**Comment by rockwotj** [comment]
golangci-lint failed due to out of space in GHA 😞

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc:24]: > Also, both processors in the PR need adding to the nav.adoc to show up in the left-hand navigation.

Where is nav.adoc? I don't know where that is

Comment (asimms41) [reply docs/modules/components/pages/processors/ollama_chat.adoc:24]: I'll need to do this in the other repo but it's just a reminder because needs to go out at the same time. This caught me out on Friday.

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (asimms41) [reply docs/modules/components/pages/processors/ollama_chat.adoc]: You could say this?

The address of the Ollama server to use. The processor starts and runs a local Ollama server unless you specify the address of a local or remote server.

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (asimms41) [reply docs/modules/components/pages/processors/ollama_chat.adoc]: Is there any more details we need to give, otherwise this will just be a duplication of the short description at the top of the page?

File: internal/impl/ollama/chat_processor.go
Comment (asimms41) [inline internal/impl/ollama/chat_processor.go:41]:
```suggestion
		Description(`This processor ends prompts to your chosen Ollama large language model (LLM) and generates text from the responses, using the Ollama API.
```

File: internal/impl/ollama/chat_processor.go
Comment (asimms41) [inline internal/impl/ollama/chat_processor.go:47]:
```suggestion
				Description("The address of the Ollama server to use. By default, a local Ollama server starts and runs unless you specify the address of a local or remote server.").
```

File: internal/impl/ollama/chat_processor.go
Comment (asimms41) [inline internal/impl/ollama/chat_processor.go:58]: The sentence 'This field supports....' is duplicated in the docs.

File: internal/impl/ollama/embeddings_processor.go
Comment (asimms41) [inline internal/impl/ollama/embeddings_processor.go:40]:
```suggestion
		Description(`This processor sends text to your chosen Ollama large language model (LLM) and creates vector embeddings, using the Ollama API. Vector embeddings are long arrays of numbers that represent values or objects, in this case text. 
```

File: internal/impl/ollama/embeddings_processor.go
Comment (asimms41) [inline internal/impl/ollama/embeddings_processor.go:47]:
```suggestion
				Description("The address of the Ollama server to use. By default, a local Ollama server starts and runs unless you specify the address of a local or remote server.").
```

File: internal/impl/ollama/embeddings_processor.go
Comment (asimms41) [inline internal/impl/ollama/embeddings_processor.go:43]:
```suggestion
For more information, see the https://ollama.com/[Ollama website^]`).
```

**Review by asimms41 — CHANGES_REQUESTED** [review]
Just minor changes.

File: internal/impl/ollama/chat_processor.go
Comment (rockwotj) [reply internal/impl/ollama/chat_processor.go]: Fixed! Thanks

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc]: Done!

File: docs/modules/components/pages/processors/ollama_chat.adoc
Comment (rockwotj) [reply docs/modules/components/pages/processors/ollama_chat.adoc]: I added more details

**Review by Jeffail — APPROVED** [review]
LGTM, thanks @rockwotj.

**Pull request merged**
